bundles/org.openhab.ui.basic/web-src/smarthome.js

```diff
@@ -1491,6 +1511,79 @@
 		_t.input.addEventListener("change", onChange);
 	}
 
+	/* class ControlInput extends Control */
+	function ControlInput(parentNode) {
+		Control.call(this, parentNode);
+
+		var
+			_t = this;
+
+		_t.input = _t.parentNode.querySelector("input[type=text]");
+		_t.itemType = _t.parentNode.getAttribute(o.itemTypeAttribute);
+		_t.verify = undefined;
+
+		var
+			lastValue = _t.input.value,
+			dotSeparatorPattern = /^-?(([0-9]{1,3}(,[0-9]{3})*)|([0-9]*))?(\.[0-9]+)?$/,
+			commaSeparatorPattern = /^-?(([0-9]{1,3}(\.[0-9]{3})*)|([0-9]*))?(,[0-9]+)?$/;
+
+		function onChange() {
+			var
+				changeValue = _t.input.value,
+				changed = true;
+			if (_t.itemType === "Number") {
+				changeValue = changeValue.trim();
+				var valueArray = changeValue.split(" ");
+				changeValue = valueArray[0];
+				if (changeValue === "") {
+					changed = false;
+				} else {
+					changeValue = changeValue.replace(/^\+/, "");
+					if (commaSeparatorPattern.test(changeValue) && !dotSeparatorPattern.test(changeValue)) {
+						changeValue = changeValue.replace(/\./g, "").replace(",", ".");
+					}
+					if (valueArray.length > 1) {
+						changeValue = changeValue + " " + valueArray[1];
+					}
+				}
+			}
+
+			if (!changed) {
+				_t.setValuePrivate(lastValue);
+			} else {
+				_t.parentNode.dispatchEvent(createEvent("control-change", {
+					item: _t.item,
+					value: changeValue
+				}));
+				// We don't know if the sent value is a valid command and will update the item state.
+				// If we don't receive an update in 1s, revert to the previous value.
+				_t.verify = new WaitingTimer(function() {
+					_t.setValuePrivate(lastValue);
+				}, 1000);
+				_t.verify.wait();
+			}
+		}
+
+		_t.setValuePrivate = function(value) {
+			if (_t.verify) {
+				_t.verify.cancel();
+			}
+			_t.input.value = value;
+			lastValue = value;
+		};
+
+		_t.setValueColor = function(color) {
+			_t.input.style.color = color;
```

Comment:
You should also update the color of the input label here in case the state is NULL/UNDEF but you receive an update of value color.

Reply:
No, because it will not be shown anymore when there is a value. The input label represents the placeholder.

Reply:
> And as you can never replace a value by a NULL or UNDEF from the input element, it will never be required.

This will never be triggered by the UI but can be triggered by a binding or a rule for example.
We can even try it with a rule that will set the item state to UNDEF for example.

Reply:
But then it will be done in the Java code that updates the widget snippet, not the javascript code.

Reply:
When the page is already opened, the Java code is no more used, the JavaScript code (front-end) will then receive directly sitemap SSE events from the OH server and use them to update widget values and colors.

Reply:
The labelcolor and iconcolor for the row is updated in the parent class (Control), so no need to do it here.
The label tag for the input element is the placeholder, and I prefer not to give that a specific color, but a neutral representation showing it is not set. If a rule changes it to UNDEF or NULL through a postUpdate, it should be the placeholder (label) that is shown, and not the value anymore, so in the neutral color. I tested this, but I did need to refresh the BasicUI to see the update to NULL or UNDEF. Otherwise it did not update. It looks like the SSE subscription did not pass through a state update to NULL or UNDEF. A sendCommand will not work as it does not accept NULL or UNDEF as valid commands. Other widgets have the same issue, proof in point, a text widget with the same value also did not update to NULL or UNDEF.